File: python/discrete_hmm.py
"""
:Module: discrete_hmm
:Synopsis: DiscreteHMM class implementation.
:Author: DOHMATOB Elvis Dopgima
"""

# TODO: 
#     - Refactor the code
#     - Documentation with sphinx
#     - Profiling 
#     - Cython optimization

__all__ = ['DiscreteHMM']

from numpy import array, sum, nonzero, zeros, abs, inf, log, exp, arange, argmax, copy
from numpy.random import shuffle
import unittest
import os

import sys
sys.path.append(os.path.dirname(os.path.dirname(sys.argv[0])) + "/python")
from probability import normalize_probabilities, almost_uniform_matrix, almost_uniform_vector, is_stochastic
from entropy_map import entropic_reestimate
from convergence import check_converged

def chopper(filename):
    ifh = open(filename)
    while 0x1:
        line = ifh.readline()
        if not line:
            break
        yield [int(item) for item in line.rstrip('\r\n').split(' ')]
    # ifh.close()


class DiscreteHMM:
    _REALTYPE = 'float64'

    def __init__(self, nstates=None, nsymbols=None, transition=None, emission=None, pi=None):
        assert not ((nstates is None and transition is None and pi is None) or (nsymbols is None and emission is None))

        if not transition is None:
            assert transition.ndim == 2
            if not nstates is None:
                assert nstates == transition.shape[0]
            else:
                nstates = transition.shape[0]
            if not emission is None:
                assert transition.shape[0] == emission.shape[0]
            if not pi is None:
                assert transition.shape[0] == pi.size
            
        if not emission is None:
            assert emission.ndim == 2
            if not nsymbols is None:
                assert nsymbols == emission.shape[1]
            else:
                nsymbols = emission.shape[1]
            if not nstates is None:
                assert nstates == emission.shape[0]
            else:
                nstates == emission.shape[0]
            if not pi is None:
                assert emission.shape[0] == pi.size

        if not pi is None:
            assert pi.ndim == 1
            if not nstates is None:
                assert nstates == pi.size
            else:
                nstates = pi.size

        if transition is None:
            transition = almost_uniform_matrix(nstates)

        if emission is None:
            emission = almost_uniform_matrix(nstates, nsymbols)

        if pi is None:
            pi = almost_uniform_vector(nstates)

        self.set_transition(transition)
        self.set_emission(emission)
        self.set_pi(pi)

    def set_transition(self, transition):
        assert is_stochastic(transition)
        self._transition = normalize_probabilities(transition, dtype=self._REALTYPE)

    def set_emission(self, emission):
        assert is_stochastic(emission)
        self._emission = normalize_probabilities(emission, dtype=self._REALTYPE)

    def set_pi(self, pi):
        assert is_stochastic(pi)
        self._pi = normalize_probabilities(pi, dtype=self._REALTYPE)

    def get_transition(self):
        return self._transition

    def get_emission(self):
        return self._emission

    def get_pi(self):
        return self._pi

    def get_nstates(self):
        return self._pi.size

    def get_nsymbols(self):
        return self._emission.shape[1]

    def forward_backward(self, obseq):
        # local variables
        T = len(obseq)
        scalers = zeros(T);
        alphatilde = zeros((T, self.get_nstates()))
        alphahat = zeros((T, self.get_nstates()))
        betatilde = zeros((T, self.get_nstates()))
        betahat = zeros((T, self.get_nstates()))
        gammahat = zeros((T, self.get_nstates()))
        epsilonhat = zeros((T-1, self.get_nstates(), self.get_nstates()))
        
        # compute forward (alpha) variables
        for time in xrange(T):
            if time == 0:
                alphatilde[time,...] = self._pi*self._emission[...,obseq[time]]
            else:
                alphatilde[time,...] = sum(array([alphahat[time-1,...],]*self.get_nstates()).T*self._transition, axis=0)*self._emission[...,obseq[time]]
            scalers[time] = 1/alphatilde[time,...].sum()
            alphahat[time,...] = scalers[time]*alphatilde[time,...]
            
        # compute backward (beta) variables
        for time in xrange(T):
            reverse_time = T-1-time
            if reverse_time == T-1:
                betatilde[reverse_time,...] = 1
            else:
                betatilde[reverse_time,...] = sum(self._transition*array([self._emission[...,obseq[reverse_time+1]]*betahat[reverse_time+1,...],]*self.get_nstates()),axis=1)
            betahat[reverse_time,...] = scalers[reverse_time]*betatilde[reverse_time,...]

        # compute epsilon and gamma terms
        gammahat = alphahat*betahat/array([scalers,]*self.get_nstates()).T
        for time in xrange(T):
            if time < T-1:
                epsilonhat[time,...,...] = array([alphahat[time,...],]*self.get_nstates()).T*self._transition*array([self._emission[...,obseq[time+1]]*betahat[time+1,...],]*self.get_nstates())

        # compute likelihood
        likelihood = -sum(log(scalers))

        # render results
        return {"alphahat":alphahat, "betahat":betahat, "gammahat":gammahat, "epsilonhat":epsilonhat, "likelihood":likelihood}

    def compute_ess(self, obseqs):
        # likelihood of learned model
        likelihood = 0 

        # expected sufficient statistics
        expected_num_transitions = zeros(self._transition.shape) 
        expected_num_emissions = zeros(self._emission.shape)
        expected_num_occurrences = zeros(self._pi.shape)

        # we'll now process all the observations and learn a new model
        for obseq in obseqs:
            # run Forwad-Backward algorithm
            fb = self.forward_backward(obseq)

            # update likelihood
            likelihood += fb.get('likelihood')

            # for each state i, update the expected number of occurences of state i
            expected_num_occurrences += fb.get('gammahat')[0,...]

            # for state pir (i, j), update the expected number of transitions i -> j
            expected_num_transitions += fb.get('epsilonhat').sum(axis=0)

            # for each state-symbol pair (i, s), update the expected number of emissions i -> s
            for j in xrange(self.get_nsymbols()):
                expected_num_emissions[...,j] += fb.get('gammahat')[nonzero(array(obseq)==j)[0],...].sum(axis=0) 

        # render results
        return {'expected_num_transitions':expected_num_transitions, \
                    'expected_num_emissions':expected_num_emissions, \
                    'expected_num_occurrences':expected_num_occurrences, 'likelihood':likelihood}
                
    def do_baumwelch(self, obseqs):
        # compute expected sufficient statistics
        ess = self.compute_ess(obseqs)

        # re-estimated model parameters using Baum-Welch MLE
        pi = normalize_probabilities(ess.get('expected_num_occurrences'))
        transition = normalize_probabilities(ess.get('expected_num_transitions'))
        emission = normalize_probabilities(ess.get('expected_num_emissions'))
        likelihood = ess.get('likelihood')

        # render results
        return {'transition':transition, 'emission':emission, 'pi':pi, 'likelihood':ess.get('likelihood')}

    def do_brand(self, obseqs):
        # compute expected sufficient statistics
        ess = self.compute_ess(obseqs)

        # re-estimate model parameters using Matthiew Brand's entropic method
        transition = zeros(self._transition.shape)
        emission = zeros(self._emission.shape)
        pi, _, _ = entropic_reestimate(ess.get('expected_num_occurrences'), theta=copy(self._pi))
        for state in xrange(self.get_nstates()):
            transition[state,...], _, _ = entropic_reestimate(ess.get('expected_num_transitions')[state,...], \
                                                                  theta=copy(self._transition[state,...]))
            emission[state,...], _, _ = entropic_reestimate(ess.get('expected_num_emissions')[state,...], \
                                                                theta=copy(self._emission[state,...]))

        # trim uninformative transitions
        denom = copy(transition)
        denom[nonzero(transition==0)] = 1
        gradient = ess.get('expected_num_transitions')/denom
        transition[nonzero(transition < exp(-gradient))] = 0
        
        # trim uninfirmative emissions
        denom = copy(emission)
        denom[nonzero(emission==0)] = 1
        gradient = ess.get('expected_num_emissions')/denom
        emission[nonzero(emission < exp(-gradient))] = 0

        # render results
        return {'transition':transition, 'emission':emission, 'pi':pi, 'likelihood':ess.get('likelihood')}

    def learn(self,
              obseqs,
              tol=1e-6,
              miniter=25,
              maxiter=100,
              method='baumwelch'):
        likelihood = -inf
        
        iteration = 0        
        while True:
            print "Iteration:", iteration
            print "Current model:", str(self)

            # budget exhausted ?
            if maxiter <= iteration:
                print "Model did not converge after %d iterations (tolerance was set to %s)."%(iteration,tol)
                break

            # learn new model
            if method == 'baumwelch':
                result = self.do_baumwelch(obseqs)
            elif method == 'brand':
                result = self.do_brand(obseqs)
            else:
                raise RuntimeError, "Unsupported parameter-estimation method: %s"%method

            print "Model likelihood:", result.get('likelihood')

            # converged ?
            if result.get('likelihood') == 0 and miniter <= iteration:
                print "Model converged (to global optimum) after %d iterations."%iteration
                break

            # update model likelihood
            converged, increased, relative_error = check_converged(likelihood, result.get('likelihood'), tol=tol)
            likelihood = result.get('likelihood')
            if method == 'baumwelch':
                assert increased # if this fails, then somethx is terribly wrong with the do_baumwelch code!!!

            print "Relative error in model likelihood over last iteration:", relative_error
            print 

            # converged ?
            if converged and miniter <= iteration:
                print "Model converged after %d iterations (tolerance was set to %s)."%(iteration,tol)
                break

            # update model
            self.set_transition(result.get('transition'))
            self.set_emission(result.get('emission'))
            self.set_pi(result.get('pi'))

            # proceed with next iteration
            iteration += 1

    def viterbi_decode(self, obseq):
        T = len(obseq)
        hiddenseq = zeros(T);
        delta = zeros((T,self.get_nstates())) # likelihoods of all paths (incomplete) presently under consideration
        phi = zeros((T,self.get_nstates()))

        # take logs of probability terms so we don't suffer underflow, etc.
        logtransition = log(self._transition)
        logemission = log(self._emission)
        logpi = log(self._pi)

        # compute stuff for time = 0
        delta[0,...] = logpi + logemission[...,obseq[0]]
        phi[0,...] = arange(self.get_nstates())
        
        # run Viterbi decoder proper
        for time in xrange(1,T):
            for j in xrange(self.get_nstates()):
                tmp = delta[time-1,...] + logtransition[...,j]
                x = argmax(tmp)
                delta[time,j] = tmp[x] + logemission[j,obseq[time]]
                phi[time,j] = x

        # set last node of optimal path
        hiddenseq[T-1] = argmax(delta[T-1,...])
        likelihood = delta[T-1,hiddenseq[T-1]]
        
        # backtrack
        for time in xrange(T-1):
            reverse_time = T-2-time
            hiddenseq[reverse_time] = phi[reverse_time+1,hiddenseq[reverse_time+1]]

        # render results
        return {'states':list(hiddenseq), 'likelihood':likelihood}

    def __dict__(self):
        return {'transition':self._transition, 'emission':self._emission, 'pi':self._pi}

    def __str__(self):
        return '%s'%self.__dict__
    

class TestDiscreteHMM(unittest.TestCase):
    # def test_constructor_with_matrices(self):
    #     dhmm = DiscreteHMM(transition=mat([[0.6, 0.4],[0.3, 0.7]]), emission=mat([[1, 1],[2, 3]]), pi=array([5,6]))
    #     self.assertTrue(is_stochastic(dhmm.get_transition()))
    #     self.assertTrue(is_stochastic(dhmm.get_emission()))
    #     self.assertTrue(is_stochastic(dhmm.get_pi()))
    #     self.assertEqual(dhmm.get_nstates(), 2)
    #     self.assertEqual(dhmm.get_nsymbols(), 2)
        
    # def test_constructor_with_sizes(self):
    #     dhmm = DiscreteHMM(pi=array([0.6, 0.4]), nsymbols=3)
    #     self.assertTrue(is_stochastic(dhmm.get_transition()))
    #     self.assertTrue(is_stochastic(dhmm.get_emission()))
    #     self.assertTrue(is_stochastic(dhmm.get_pi()))
    #     self.assertEqual(dhmm.get_nstates(), 2)
    #     self.assertEqual(dhmm.get_nsymbols(), 3)

    def test_learn(self):
        dhmm = DiscreteHMM(transition=array([[0.5,0.5],[0.5,0.5]]),
                           emission=array([[1,1,1],[1,1,1]], dtype='float64'),
                           pi=array([0.6,0.4]),
                           )
          
        result = dhmm.do_baumwelch([[0,1,0,0,1,1]])
        print result
        dhmm.set_pi(result.get('pi'))
        dhmm.set_transition(result.get('transition'))
        dhmm.set_emission(result.get('emission'))

        result = dhmm.forward_backward([0,1,0,0,1,1])
        print result
        

    # # def test_viterbi(self):
    # #     dhmm = DiscreteHMM(transition=array([[0.4,0.6],[0.6,0.4]]),
    # #                        emission=array([[0.67,0.33],[0.41,0.59]]),
    # #                        pi=array([0.75,0.25]),
    # #                        )

    # #     self.assertEqual(dhmm.viterbi_decode([1,0,0,1,0,1]).get('states'), [0,1,0,1,0,1])
              
def main():
    dhmm = DiscreteHMM(nstates=2, nsymbols=26)
    lessons = list(chopper('data/corpus_words.dat'))
    shuffle(lessons)
    dhmm.learn(lessons[:500], method='brand', maxiter=100)
    
    print 
    print 'Viterbi classification of 26 symbols (cf. letters of the English alphabet):'
    clusters = dict()
    for i in xrange(26):
        letter = chr(ord('A')+i)
        class_label = dhmm.viterbi_decode([i]).get('states')[0]
        if i == 0:
            # 'a' is a vowel and 'a' is not a consonant
            clusters[class_label] = 'vowel'
            clusters[1-class_label] = 'consonant'
        print '\t%s is a %s'%(letter,clusters[class_label])

if __name__ == '__main__':
    if 'TEST' in os.environ:
        unittest.main()
    else:
        main()

File: src/DiscreteHMM.cxx
/*!
  \file DiscreteHMM.cxx
  \brief Implementation of DiscreteHMM class.
  \author DOP (dohmatob elvis dopgima)
*/

#include "DiscreteHMM.h"
#include "HMMUtils.h"
#include "HMMPathType.h"
#include <boost/assert.hpp>
#include <boost/numeric/ublas/matrix_proxy.hpp> // for constructing matrix proxies like, row, column, etc.
#include "ProgressBar.h"

namespace ublas = boost::numeric::ublas;

HiddenMarkovModels::DiscreteHMM::DiscreteHMM(HiddenMarkovModels::RealMatrixType& transition, 
					     HiddenMarkovModels::RealMatrixType& emission, 
					     HiddenMarkovModels::RealVectorType& pi)
{
  // sanity checks on dimensions of matrices
  BOOST_ASSERT(transition.size1() == transition.size2()); // transition matrix should be square
  BOOST_ASSERT(pi.size() == transition.size1()); // transition matrix should have as many rows as there are hidden states in the model
  BOOST_ASSERT(transition.size1() == emission.size1()); // transition and emission matrices should have same number of rows

  // initialize model parameters proper
  set_transition(transition);
  set_emission(emission);
  set_pi(pi);
}

void HiddenMarkovModels::DiscreteHMM::set_transition(const HiddenMarkovModels::RealMatrixType& transition)
{
  BOOST_ASSERT(HiddenMarkovModels::is_stochastic_matrix(transition));

  _transition = transition;
}

void HiddenMarkovModels::DiscreteHMM::set_emission(const HiddenMarkovModels::RealMatrixType& emission)
{
  BOOST_ASSERT(HiddenMarkovModels::is_stochastic_matrix(emission));

  _emission = emission;
}

void HiddenMarkovModels::DiscreteHMM::set_pi(const HiddenMarkovModels::RealVectorType& pi)
{
  BOOST_ASSERT(HiddenMarkovModels::is_stochastic_vector(pi));

  _pi = pi;
}

int HiddenMarkovModels::DiscreteHMM::get_nstates() const
{
  return _pi.size();
}

int HiddenMarkovModels::DiscreteHMM::get_nsymbols() const
{
  return _emission.size2();
}

const HiddenMarkovModels::RealMatrixType& HiddenMarkovModels::DiscreteHMM::get_transition() const
{
  return _transition;
}

const HiddenMarkovModels::RealMatrixType& HiddenMarkovModels::DiscreteHMM::get_emission() const
{
  return _emission;
}

const HiddenMarkovModels::RealVectorType& HiddenMarkovModels::DiscreteHMM::get_pi() const
{
  return _pi;
}

bool HiddenMarkovModels::DiscreteHMM::is_symbol(unsigned int i) const
{
  return 0 <= i && i < get_nsymbols();
}

HiddenMarkovModels::HMMPathType HiddenMarkovModels::DiscreteHMM::viterbi(const HiddenMarkovModels::ObservationSequenceType& obseq)
{
  // variables
  int T = obseq.size();
  HiddenMarkovModels::ObservationSequenceType hiddenseq(T); // optimal path (sequence of hidden states that generated observed trace)
  HiddenMarkovModels::RealType likelihood;
  HiddenMarkovModels::RealMatrixType delta(T, get_nstates());
  ublas::matrix<int> phi(T, get_nstates());

  // take logarithms, so we don't suffer underflow!
  HiddenMarkovModels::RealMatrixType logtransition = HiddenMarkovModels::mlog(_transition);
  HiddenMarkovModels::RealMatrixType logemission = HiddenMarkovModels::mlog(_emission);
  HiddenMarkovModels::RealVectorType logpi = HiddenMarkovModels::vlog(_pi);

  // compute stuff for time = 0
  BOOST_ASSERT(is_symbol(obseq[0]));
  row(delta,0) = logpi + column(logemission, obseq[0]);
  for(int j = 0; j < get_nstates(); j++)
    {
      phi(0,j) = j;
    }

  // run viterbi proper
  for(int time = 1; time < T; time++)
    {
      for(int j = 0; j < get_nstates(); j++)
	{
	  HiddenMarkovModels::RealVectorType tmp = row(delta, time-1)+column(logtransition, j);
	  boost::tuple<int,
		       HiddenMarkovModels::RealType
		       > x = HiddenMarkovModels::argmax(tmp);
	  BOOST_ASSERT(is_symbol(obseq[time]));
	  delta(time, j) = boost::get<1>(x)+logemission(j, obseq[time]);
	  phi(time, j) = boost::get<0>(x);
	}
    }

  // set last node of optimal path
  HiddenMarkovModels::RealVectorType tmp = row(delta, T-1);
  boost::tuple<int,
	       HiddenMarkovModels::RealType
	       > x = HiddenMarkovModels::argmax(tmp);
  likelihood = boost::get<1>(x);
  int state = boost::get<0>(x);
  hiddenseq[T-1] = state;

  // backtrack
  for(int time = T-2; time >= 0; time--)
    {
      state = phi(time+1, state);
      hiddenseq[time] = state;
    }

  return HiddenMarkovModels::HMMPathType(hiddenseq, likelihood);
}

boost::tuple<HiddenMarkovModels::RealMatrixType, // alpha-hat
	     HiddenMarkovModels::RealMatrixType, // beta-hat
	     HiddenMarkovModels::RealMatrixType, // gamma-hat
	     boost::multi_array< HiddenMarkovModels::RealType,3 >, // epsilon-hat
	     HiddenMarkovModels::RealType // likelihood
	     > HiddenMarkovModels::DiscreteHMM::forward_backward(const HiddenMarkovModels::ObservationSequenceType &obseq)
{
  // variables
  unsigned int T = obseq.size();
  HiddenMarkovModels::RealVectorType scalers(T); // these things will prevent underflow, etc.
  HiddenMarkovModels::RealMatrixType alphatilde(T, get_nstates());
  HiddenMarkovModels::RealMatrixType alphahat(T, get_nstates()); // forward variables
  HiddenMarkovModels::RealMatrixType betatilde(T, get_nstates()); 
  HiddenMarkovModels::RealMatrixType betahat(T, get_nstates()); // backward variables
  HiddenMarkovModels::RealMatrixType gammahat(T, get_nstates());
  HiddenMarkovModels::RealVectorType tmp(get_nstates());
  boost::multi_array<HiddenMarkovModels::RealType,3> epsilonhat(boost::extents[T-1][get_nstates()][get_nstates()]);
  HiddenMarkovModels::RealType likelihood;

  // compute forward (alpha) variables
  for(int time = 0; time < T; time++)
    {
      BOOST_ASSERT(is_symbol(obseq[time]));
      if(time == 0)
	{
	  row(alphatilde, time) = element_prod(_pi, column(_emission, obseq[time]));
	}
      else
	{
	  for(int i = 0; i < get_nstates(); i++)
	    {
	      alphatilde(time, i) = sum(element_prod(row(alphahat, time-1), column(_transition,i))) * _emission(i, obseq[time]);
	    }
	}
      scalers[time] = 1/sum(row(alphatilde, time)); 
      row(alphahat, time) = scalers[time]*row(alphatilde, time);
    }

  // compute backward (beta) parameters
  for(int time = T-1; time >= 0; time--)
    {
      if (time == T-1)
	{
	  row(betatilde, time) = ublas::scalar_vector<HiddenMarkovModels::RealType>(get_nstates(), 1);
	}
      else
	{
	  for(int i = 0; i < get_nstates(); i++)
	    {
	      betatilde(time, i) = sum(element_prod(element_prod(row(_transition, i), column(_emission, obseq[time+1])), row(betahat, time+1)));
	    }
	}
      row(betahat, time) = scalers[time]*row(betatilde, time);
    }

  // compute epsilon and gamma terms
  for(int time = 0; time < T; time++)
    {
      row(gammahat, time) = element_prod(row(alphahat, time), row(betahat, time))/scalers[time];
      if (time < T-1)
	{
	  for(int i = 0; i < get_nstates(); i++)
	    {
	      for(int j = 0; j < get_nstates(); j++)
		{
		  epsilonhat[time][i][j] = alphahat(time, i)*_transition(i, j)*_emission(j, obseq[time+1])*betahat(time+1, j);
		}
	    }
	}
    }

  // compute likelihood
  likelihood = -1*sum(HiddenMarkovModels::vlog(scalers));

  return boost::make_tuple(alphahat, betahat, gammahat, epsilonhat, likelihood);
}

boost::tuple<HiddenMarkovModels::DiscreteHMM, // the learned model
	     HiddenMarkovModels::RealType // the likelihood of the sequences of observations, under this new model
	     > HiddenMarkovModels::DiscreteHMM::learn(const std::vector< HiddenMarkovModels::ObservationSequenceType > &obseqs)
{
  // local typedefs
  typedef boost::multi_array<HiddenMarkovModels::RealType,3> floats3D;
  typedef floats3D::index_range range;
  floats3D::index_gen indices;

  // local variables
  HiddenMarkovModels::RealType likelihood = 0;
  HiddenMarkovModels::RealMatrixType transition(get_nstates(), get_nstates()); // transition matrix for learned model
  HiddenMarkovModels::RealMatrixType emission(get_nstates(), get_nsymbols()); // emission matrix for learned model
  HiddenMarkovModels::RealVectorType pi(get_nstates()); // initial distribution for learned model
  HiddenMarkovModels::RealMatrixType u(get_nstates(),get_nstates());
  HiddenMarkovModels::RealMatrixType w(get_nstates(),get_nsymbols());
  HiddenMarkovModels::RealVectorType v(get_nstates());
  HiddenMarkovModels::RealVectorType x(get_nstates());
  
  // initializations
  likelihood = 0;
  pi = x = v = 0*_pi;
  u = ublas::zero_matrix<HiddenMarkovModels::RealType >(u.size1(), u.size2());
  w = ublas::zero_matrix<HiddenMarkovModels::RealType >(w.size1(), w.size2());

  // process all observations (assumed to be independent!)
  for(int k = 0; k < obseqs.size(); k++)
    {
      // length of observation
      int T = obseqs[k].size();

      // run Forward-Backward
      boost::tuple<HiddenMarkovModels::RealMatrixType, // alpha-hat
		   HiddenMarkovModels::RealMatrixType, // beta-hat
		   HiddenMarkovModels::RealMatrixType, // gamma-hat
		   boost::multi_array<HiddenMarkovModels::RealType, 3>, // epsilon-hat
		   HiddenMarkovModels::RealType // likelihood
		   >  fb = forward_backward(obseqs[k]);
      HiddenMarkovModels::RealMatrixType alphahat = boost::get<0>(fb);
      HiddenMarkovModels::RealMatrixType betahat = boost::get<1>(fb);
      HiddenMarkovModels::RealMatrixType gammahat = boost::get<2>(fb);
      boost::multi_array<HiddenMarkovModels::RealType, 3> epsilonhat = boost::get<3>(fb);

      // update likelihood
      likelihood += boost::get<4>(fb);

      // update pi
      pi += row(gammahat, 0);

      // update auxiliary tensors
      for(int i = 0; i < get_nstates(); i++)
	{
	  x[i] += sum(column(gammahat, i));
	  v[i] += sum(column(gammahat, i)) - gammahat(T-1, i);
	  for(int j = 0; j < get_nstates(); j++)
	    {
	      for(int time = 0; time < T-1; time++)
		{
		      u(i,j) += epsilonhat[time][i][j];
		}
	    }
	  for(int j = 0; j < get_nsymbols(); j++)
	    {
	      for(int time = 0; time < T; time++)
		{
		  if (obseqs[k][time] == j)
		    {
		      w(i,j) += gammahat(time, i);
		    }
		}
	    }
	}
    }

  // compute learned model parameters
  pi /= obseqs.size(); // XXX if this normalization doesn't make pi sum to 1, then the something is horribly wrong with the code!!!
  for(int i = 0; i < get_nstates(); i++)
    {
      row(transition, i) = row(u, i)/v[i];
      row(emission, i) = row(w, i)/x[i];
    }

  return boost::make_tuple(HiddenMarkovModels::DiscreteHMM(transition, emission, pi), likelihood);
}

HiddenMarkovModels::RealType HiddenMarkovModels::DiscreteHMM::baumwelch(const std::vector< HiddenMarkovModels::ObservationSequenceType > &obseqs,
									 HiddenMarkovModels::RealType tolerance,
									 int maxiter
									 )
{
  // intializations
  int iteration = 0;
  HiddenMarkovModels::RealType likelihood = -1*std::numeric_limits< HiddenMarkovModels::RealType >::max(); // minus infinity
  HiddenMarkovModels::RealType relative_gain = 0;
  BeautifulThings::ProgressBar progressbar;

  std::cout << "Running Baum-Welch algorithm on a pool of " << obseqs.size() << " observation sequence(s) .." << std::endl;

  // main loop
  while (true)
    {
      // update progress bar
      progressbar.update(iteration, maxiter);

      // done ?
      if (0 <= maxiter && maxiter <= iteration)
	{
	  std::cout << std::endl << "Model did not converge after " << iteration << " iteration(s)." << std::endl;
	  break;
	}

      // learn
      boost::tuple<HiddenMarkovModels::DiscreteHMM,
		   HiddenMarkovModels::RealType
		   > learned_hmm = learn(obseqs);
      DiscreteHMM new_hmm = boost::get<0>(learned_hmm);
      RealType new_likelihood = boost::get<1>(learned_hmm); 
   
      // converged ?
      if (new_likelihood == 0)
	{
	  progressbar.update(maxiter, maxiter);
	  std::cout << std::endl << "Converged (to global optimum!) after " << iteration << " iteration(s)." << std::endl;
	  break;
	}

      // update this model
      relative_gain = (new_likelihood - likelihood)/abs(new_likelihood);
      BOOST_ASSERT(relative_gain >= 0); // if this fails, then something is terribly wrong with the implementation!
      set_transition(new_hmm.get_transition());
      set_emission(new_hmm.get_emission());
      set_pi(new_hmm.get_pi());

      // update likehood
      likelihood = new_likelihood;

      // converged ?
      if (relative_gain < tolerance)
	{
	  progressbar.update(maxiter, maxiter);
	  std::cout << std::endl << "Converged after " << iteration << " iteration(s) (tolerance was set to " << tolerance << ")." << std::endl;
	  break;
	}

      // proceed with next iteration
      iteration++;
    }

  std::cout << std::endl << "Final model: " << std::endl << *this << std::endl;
  std::cout << std::endl << "Likelihood: " << likelihood << std::endl;
  std::cout << "Relative gain in likelihood over last iteration: " << relative_gain << std::endl << std::endl;
  
  return likelihood;
}
